File: scripts/benchmark.mjs
#!/usr/bin/env node
import { writeFileSync } from 'fs';
import { encode } from '../src/encoder.js';
import { tokenize, loadTokenizer } from '../src/tokenizer/index.js';
import { encode as gptEncode } from 'gpt-tokenizer';
import llamaTok from 'llama-tokenizer-js/llama-tokenizer.js';

const tok = llamaTok;
const vocab = loadTokenizer();

const TESTS = [
  ['Common English', 'the quick brown fox jumps over the lazy dog this is a test of the emergency broadcast system how now brown cow the rain in spain stays mainly on the plain'],
  ['Code', 'const app = express(); app.get("/api/tasks", async (req, res) => { const tasks = await db.query("SELECT * FROM tasks"); res.json(tasks); });'],
  ['SQL', 'SELECT * FROM users WHERE id=42 AND name LIKE \'%test%\' ORDER BY created_at DESC; INSERT INTO table_name (a,b,c) VALUES (1,\'x\',true); UPDATE users SET name=\'abc\', score=99 WHERE id=7; DELETE FROM users WHERE id>1000; CREATE TABLE test(id INT PRIMARY KEY,name VARCHAR(255));'],
  ['API', '{"status": "success", "data": {"users": [{"id": 1, "name": "John", "email": "john@example.com"}, {"id": 2, "name": "Jane", "email": "jane@example.com"}], "total": 2, "page": 1, "per_page": 10}}'],
  ['Shell', '$ echo "Hello, World!"; printf \'%s\\n\' "$HOME"; ls -la /tmp | grep ".log" && cat file.txt; sudo -n true || echo "no sudo"; python3 -c \'print("test")\'; node -e "console.log(42)"; git status --short; git add . && git commit -m "test"; npm run build && npm start'],
  ['Markdown', '# README.md ## Test Project ### Features - fast - simple - random ### Code `npm install && npm run dev` **bold** *italic* [link](https://example.com) > quote --- ### End'],
  ['Paths', 'C:\\Users\\Test\\file.txt D:\\Games\\MC\\server.exe /usr/bin/bash ~/.config/hypr/hyprland.conf ../../src/main.js ./build/output.log https://example.com/?a=1&b=2 ftp://x@y.z:21/path git@host:user/repo.git user@example.com test+tag@example.org'],
  ['Prompt', 'aicl is Goated BTW, and this can reduce tokens very vary fast'],
];

const rows = TESTS.map(([name, text]) => {
  const aicl = tokenize(encode(text).output, vocab).length;
  const gpt3 = gptEncode(text, { model: 'gpt-3.5-turbo' }).length;
  const gpt4 = gptEncode(text, { model: 'gpt-4' }).length;
  const gpt4o = gptEncode(text, { model: 'gpt-4o' }).length;
  const gpt5 = gptEncode(text, { model: 'gpt-5' }).length;
  const llama = tok.encode(text).length;
  return { name, raw: text.length, aicl, gpt3, gpt4, gpt4o, gpt5, llama, win: +(gpt4o / aicl).toFixed(2) };
});

for (const r of rows) console.log(`${r.name.padEnd(16)} raw=${String(r.raw).padStart(3)} AICL=${String(r.aicl).padStart(3)} gpt3=${String(r.gpt3).padStart(3)} gpt4=${String(r.gpt4).padStart(3)} gpt4o=${String(r.gpt4o).padStart(3)} g5=${String(r.gpt5).padStart(3)} llama=${String(r.llama).padStart(3)} win=${r.win}x`);
console.log(`vocab ${vocab.numMerges} merges maxLen=${vocab.maxTokenLength}`);

// ── helpers ──
const esc = s => s.replace(/&/g,'&amp;').replace(/</g,'&lt;').replace(/>/g,'&gt;');
const W = 900;
const gradDef = `
  <defs>
    <linearGradient id="aiclGrad" x1="0" y1="0" x2="1" y2="0" gradientUnits="objectBoundingBox">
      <stop offset="0%" stop-color="#7c3aed"/><stop offset="55%" stop-color="#ec4899"/><stop offset="100%" stop-color="#f97316"/>
    </linearGradient>
    <linearGradient id="headerGrad" x1="0" y1="0" x2="1" y2="0">
      <stop offset="0%" stop-color="#7c3aed"/><stop offset="50%" stop-color="#ec4899"/><stop offset="100%" stop-color="#f97316"/>
    </linearGradient>
    <linearGradient id="vgrad" x1="0" y1="1" x2="0" y2="0">
      <stop offset="0%" stop-color="#fafafa" stop-opacity="0.95"/><stop offset="100%" stop-color="#fafafa"/>
    </linearGradient>
    <filter id="glow"><feDropShadow dx="0" dy="2" stdDeviation="3" flood-color="#ec4899" flood-opacity="0.25"/></filter>
  </defs>`;

function barRect(x, y, w, h, fill, rx = 4, extra='') {
  return `<rect x="${x}" y="${y}" width="${w}" height="${h}" rx="${rx}" fill="${fill}" ${extra}/>`;
}

// ── benchmark.svg (GPT-4o vs AICL) ──
{
  const H = 560;
  const top = 110, bottom = 410, chartH = bottom - top;
  const maxTok = Math.max(...rows.map(r => Math.max(r.gpt4o, r.aicl)), 90);
  const scale = chartH / (maxTok * 1.12);
  const cols = rows.length;
  const colW = 86, gap = (820 - cols * colW) / (cols - 1);
  const startX = 60;
  let groups = '';
  rows.forEach((r, i) => {
    const x = startX + i * (colW + gap);
    const gH = Math.round(r.gpt4o * scale);
    const aH = Math.round(r.aicl * scale);
    const gY = bottom - gH;
    const aY = bottom - aH;
    const label1 = r.name === 'Common English' ? ['Common','English'] : r.name === 'API' ? ['API','response'] : r.name === 'Normal Prompt' || r.name === 'Prompt' ? ['Prompt'] : [r.name];
    const win = r.win;
    const winColor = win >= 2 ? '#10b981' : win >= 1.2 ? '#a3e635' : '#eab308';
    groups += `
  <g>
    ${barRect(x+8, gY, 28, gH, '#27272a', 4, 'opacity="0.9"')}
    <text x="${x+22}" y="${gY-8}" text-anchor="middle" fill="#9ca3af" font-size="9" font-weight="600">${r.gpt4o}</text>
    ${barRect(x+44, aY, 28, aH, 'url(#aiclGrad)', 4, 'filter="url(#glow)"')}
    <text x="${x+58}" y="${aY-8}" text-anchor="middle" fill="#fff" font-size="9" font-weight="700">${r.aicl}</text>
    ${label1.map((l, idx) => `<text x="${x+40}" y="${440+idx*12}" text-anchor="middle" font-size="10" font-weight="600" fill="#e5e7eb">${esc(l)}</text>`).join('')}
    <text x="${x+40}" y="472" text-anchor="middle" fill="${winColor}" font-size="9" font-weight="700">${win}×</text>
  </g>`;
  });
  const gridLabels = [0.25,0.5,0.75,1].map(f=>{
    const v = Math.round(maxTok * f);
    const y = bottom - Math.round(maxTok*f*scale);
    return {v,y};
  });
  const svg = `<svg width="${W}" height="${H}" xmlns="http://www.w3.org/2000/svg">
<style>.t{font:700 19px 'Inter',system-ui,sans-serif;fill:#f9fafb;letter-spacing:-0.4px}.s{font:400 11px 'Inter',system-ui,sans-serif;fill:#9ca3af}.n{font:600 10px 'Inter',system-ui,sans-serif;fill:#e5e7eb}.a{font:400 9px 'Inter',system-ui,sans-serif;fill:#6b7280}</style>
${gradDef}
<rect width="${W}" height="${H}" rx="16" fill="#0a0a0a"/>
<rect width="${W}" height="4" rx="2" fill="url(#headerGrad)" opacity="0.95"/>
<text x="32" y="36" class="t">AICL vs GPT-4o — tokens <tspan fill="#ec4899">lower is cheaper</tspan></text>
<text x="32" y="56" class="s">2–5 PUA → 1 token · ${vocab.numMerges} merges · max 5 PUA/token · 8 tests · 131/131 pass</text>
<g transform="translate(32,68)">
  <rect x="0" y="0" width="10" height="10" rx="2" fill="#27272a"/><text x="14" y="9" class="s">GPT-4o (o200k_base)</text>
  <rect x="140" y="0" width="10" height="10" rx="2" fill="url(#aiclGrad)"/><text x="154" y="9" class="s">AICLTokenizer</text>
  <text x="260" y="9" class="s" opacity="0.6">· purple→pink→orange = VSPCODERZ</text>
</g>
<g stroke="#1f1f23" stroke-width="1">
  ${[top, top+chartH*0.25, top+chartH*0.5, top+chartH*0.75, bottom].map(y=>`<line x1="52" y1="${y}" x2="860" y2="${y}"/>`).join('')}
</g>
${gridLabels.map(({v,y})=>`<text x="44" y="${y+4}" class="a" text-anchor="end">${v}</text>`).join('')}
<text x="44" y="${top-6}" class="a" text-anchor="end">${maxTok}</text>
${groups}
<line x1="32" y1="490" x2="868" y2="490" stroke="#1f1f23"/>
<text x="32" y="510" class="s">↓ fewer tokens = cheaper · pipeline: raw → AICL (${vocab.numMerges} dict · 2–5 PUA) → BPE → tokens</text>
<text x="32" y="528" class="a">VSPCODERZ · Dream it. Design it. Craft it. · github.com/vspcoderz/aicl</text>
</svg>`;
  writeFileSync('assets/benchmark.svg', svg);
  console.log('wrote assets/benchmark.svg');
}

// ── benchmark-all.svg (6 tokenizers) ──
{
  const H = 700;
  const top = 100, bottom = 460, chartH = bottom - top;
  const maxTok = Math.max(...rows.flatMap(r=>[r.gpt3,r.gpt4,r.gpt4o,r.gpt5,r.llama,r.aicl]), 90);
  const scale = chartH / (maxTok * 1.1);
  const cols = rows.length;
  const groupW = 86, gap = (820 - cols*groupW)/(cols-1);
  const startX = 60;
  const barW = 10, barGap = 2;
  const colors = { gpt3:'#f97316', gpt4:'#eab308', gpt4o:'#06b6d4', gpt5:'#8b5cf6', llama:'#f43f5e', aicl:'url(#aiclGrad)' };
  let groups='';
  rows.forEach((r,i)=>{
    const x = startX + i*(groupW+gap);
    const vals = [r.gpt3,r.gpt4,r.gpt4o,r.gpt5,r.llama,r.aicl];
    const keys = ['gpt3','gpt4','gpt4o','gpt5','llama','aicl'];
    let bars='';
    vals.forEach((v,j)=>{
      const h = Math.round(v*scale);
      const y = bottom - h;
      const bx = x + j*(barW+barGap);
      const fill = colors[keys[j]];
      const rx = j===5?3:2;
      bars+=`<rect x="${bx}" y="${y}" width="${barW}" height="${h}" rx="${rx}" fill="${fill}" ${j===5?'filter="url(#glow)"':''} opacity="${j===5?'1':'0.88'}"/>`;
    });
    const short = r.name==='Common English'?'Common':r.name==='API'?'API':r.name;
    const win = r.win;
    groups+=`
  <g>
    ${bars}
    <text x="${x+groupW/2}" y="500" text-anchor="middle" class="n">${esc(short)}</text>
    ${r.name==='Common English'?`<text x="${x+groupW/2}" y="512" text-anchor="middle" class="n">Eng</text>`:''}
    <text x="${x+groupW/2}" y="530" text-anchor="middle" fill="#a78bfa" font-size="7" font-weight="700">${win}×</text>
  </g>`;
  });
  const svg2 = `<svg width="${W}" height="${H}" xmlns="http://www.w3.org/2000/svg">
<style>.t{font:700 18px 'Inter',system-ui,sans-serif;fill:#f9fafb}.s{font:400 11px 'Inter',system-ui,sans-serif;fill:#9ca3af}.n{font:600 9px 'Inter',system-ui,sans-serif;fill:#e5e7eb}.a{font:400 8px 'Inter',system-ui,sans-serif;fill:#6b7280}</style>
${gradDef}
<rect width="${W}" height="${H}" rx="16" fill="#0a0a0a"/>
<rect width="${W}" height="4" rx="2" fill="url(#headerGrad)" opacity="0.95"/>
<text x="32" y="34" class="t">AICL vs All Tokenizers — <tspan fill="#ec4899">8 tests</tspan></text>
<text x="32" y="54" class="s">tokens — lower is better · GPT-3/4/4o/5 + LLaMA 2 + AICL (${vocab.numMerges} merges, 2–5 PUA → 1 token)</text>
<g transform="translate(32,66)">
  <rect x="0" y="0" width="10" height="10" rx="2" fill="#f97316"/><text x="14" y="9" class="s">GPT-3</text>
  <rect x="58" y="0" width="10" height="10" rx="2" fill="#eab308"/><text x="72" y="9" class="s">GPT-4</text>
  <rect x="112" y="0" width="10" height="10" rx="2" fill="#06b6d4"/><text x="126" y="9" class="s">GPT-4o</text>
  <rect x="172" y="0" width="10" height="10" rx="2" fill="#8b5cf6"/><text x="186" y="9" class="s">GPT-5</text>
  <rect x="226" y="0" width="10" height="10" rx="2" fill="#f43f5e"/><text x="240" y="9" class="s">LLaMA2</text>
  <rect x="296" y="0" width="10" height="10" rx="2" fill="url(#aiclGrad)"/><text x="310" y="9" class="s">AICL</text>
</g>
<g stroke="#1a1a1e" stroke-width="1">
  ${[top, top+chartH*0.25, top+chartH*0.5, top+chartH*0.75, bottom].map(y=>`<line x1="52" y1="${y}" x2="860" y2="${y}"/>`).join('')}
</g>
${[0,0.25,0.5,0.75,1].map(f=>{
  const v = Math.round(maxTok*f);
  const y = bottom - Math.round(maxTok*f*scale);
  return `<text x="44" y="${y+4}" class="a" text-anchor="end">${v}</text>`;
}).join('')}
${groups}
<line x1="32" y1="560" x2="868" y2="560" stroke="#1f1f23"/>
<text x="32" y="580" class="s">AICL wins 8/8 — best: Code 4.00×, API 3.84×, Markdown 3.31×, Paths 3.13× · Lower = cheaper</text>
<text x="32" y="600" class="s">Square bars · VSPCODERZ gradient · github.com/vspcoderz/aicl · 131/131 pass · 8 tests · 6 tokenizers</text>
<text x="32" y="618" class="a">Dream it. Design it. Craft it.</text>
</svg>`;
  writeFileSync('assets/benchmark-all.svg', svg2);
  console.log('wrote assets/benchmark-all.svg');
}

File: scripts/benchmark.py
#!/usr/bin/env python3
"""Benchmark AICL Stage-2 token counts and render the two legacy SVG charts.

GPT/LLaMA counts are optional because the JavaScript implementation obtained
them from third-party packages.  This Python port remains stdlib-only; pass
``--baselines`` with the same per-test counts when those comparison charts are
needed.  AICL counts, warm-up independent, and the chart generation work with
no third-party package installed.
"""

from __future__ import annotations

import argparse
import html
import json
import statistics
import sys
import time
from pathlib import Path
from typing import Any, Mapping, Sequence

ROOT = Path(__file__).resolve().parents[1]
if str(ROOT) not in sys.path:
    sys.path.insert(0, str(ROOT))

try:
    from .eval_vocab import _load_vocab, _result_output, _tokenizer_api, vocab_summary  # noqa: E402
    from .heldout_probes import BENCH  # noqa: E402
except ImportError:
    from eval_vocab import _load_vocab, _result_output, _tokenizer_api, vocab_summary  # noqa: E402
    from heldout_probes import BENCH  # noqa: E402


def load_baselines(path: Path | None) -> dict[str, dict[str, int]]:
    if path is None:
        return {}
    raw = json.loads(path.read_text(encoding="utf-8"))
    if isinstance(raw, dict) and "rows" in raw:
        raw = raw["rows"]
    if isinstance(raw, list):
        raw = {str(item["name"]): item for item in raw}
    if not isinstance(raw, dict):
        raise ValueError("baseline file must be an object keyed by test name")
    result: dict[str, dict[str, int]] = {}
    for name, values in raw.items():
        if not isinstance(values, Mapping):
            continue
        result[str(name)] = {
            key: int(value)
            for key, value in values.items()
            if key in {"gpt3", "gpt4", "gpt4o", "gpt5", "llama"}
        }
    return result


def measure(vocab: Any, tokenize: Any, baselines: Mapping[str, Mapping[str, int]]) -> list[dict[str, Any]]:
    from aicl import encode

    rows: list[dict[str, Any]] = []
    for name, text in BENCH:
        started = time.perf_counter()
        encoded = encode(text)
        aicl = _result_output(encoded)
        ids = tokenize(aicl, vocab)
        elapsed_ms = (time.perf_counter() - started) * 1000
        row: dict[str, Any] = {
            "name": name,
            "raw": len(text),
            "aicl": len(ids),
            "ms": elapsed_ms,
            "gpt3": None,
            "gpt4": None,
            "gpt4o": None,
            "gpt5": None,
            "llama": None,
        }
        for key in ("gpt3", "gpt4", "gpt4o", "gpt5", "llama"):
            if key in baselines.get(name, {}):
                row[key] = baselines[name][key]
        if row["gpt4o"] is not None:
            row["win"] = round(row["gpt4o"] / row["aicl"], 2) if row["aicl"] else 0.0
        else:
            row["win"] = None
        rows.append(row)
    return rows


def hotpath_summary() -> dict[str, float]:
    """Measure cold startup plus warmed 10k encode/tokenize medians."""

    from aicl import encode, load_tokenizer, tokenize

    text = ("the quick brown fox jumps over the lazy dog " * 250)[:10_000]
    started = time.perf_counter()
    encoded = encode(text)["output"]
    cold_encode_ms = (time.perf_counter() - started) * 1000

    started = time.perf_counter()
    vocab = load_tokenizer()
    tokenize(encoded, vocab)
    cold_tokenize_ms = (time.perf_counter() - started) * 1000

    encode_times: list[float] = []
    tokenize_times: list[float] = []
    for _ in range(5):
        started = time.perf_counter()
        encoded = encode(text)["output"]
        encode_times.append((time.perf_counter() - started) * 1000)
        started = time.perf_counter()
        tokenize(encoded, vocab)
        tokenize_times.append((time.perf_counter() - started) * 1000)
    return {
        "cold_encode_ms": cold_encode_ms,
        "cold_tokenize_ms": cold_tokenize_ms,
        "encode_median_ms": statistics.median(encode_times),
        "tokenize_median_ms": statistics.median(tokenize_times),
    }


def _escape(value: Any) -> str:
    return html.escape(str(value), quote=True)


def _bar_chart(
    rows: Sequence[Mapping[str, Any]],
    *,
    title: str,
    subtitle: str,
    all_tokenizers: bool,
) -> str:
    width = 900
    height = 620 if all_tokenizers else 520
    top = 96
    bottom = 420 if all_tokenizers else 380
    chart_height = bottom - top
    values = [
        int(value)
        for row in rows
        for key, value in row.items()
        if key in {"aicl", "gpt3", "gpt4", "gpt4o", "gpt5", "llama"} and value is not None
    ]
    max_tokens = max([*values, 90])
    scale = chart_height / (max_tokens * (1.1 if all_tokenizers else 1.12))
    group_width = 86
    gap = (820 - len(rows) * group_width) / max(1, len(rows) - 1)
    palette = {
        "gpt3": "#71717a",
        "gpt4": "#52525b",
        "gpt4o": "#3f3f46",
        "gpt5": "#27272a",
        "llama": "#a1a1aa",
        "aicl": "#fafafa",
    }
    keys = ["gpt3", "gpt4", "gpt4o", "gpt5", "llama", "aicl"] if all_tokenizers else ["aicl"]
    bar_width = 10 if all_tokenizers else 28
    bar_gap = 2 if all_tokenizers else 0
    groups: list[str] = []
    for index, row in enumerate(rows):
        x = 60 + index * (group_width + gap)
        bars: list[str] = []
        for key_index, key in enumerate(keys):
            value = row.get(key)
            if value is None:
                continue
            bar_height = max(0, round(int(value) * scale))
            y = bottom - bar_height
            bx = x + key_index * (bar_width + bar_gap) if all_tokenizers else x + 29
            fill = palette[key]
            stroke = ' stroke="#e5e7eb" stroke-width="0.5"' if key == "aicl" else ""
            bars.append(
                f'<rect x="{bx:.1f}" y="{y:.1f}" width="{bar_width}" height="{bar_height}" '
                f'rx="{3 if key == "aicl" else 2}" fill="{fill}"{stroke}/>'
            )
        label = "Common" if row["name"] == "Common English" else "API" if row["name"] == "API" else row["name"]
        label_lines = ["Common", "English"] if row["name"] == "Common English" else ["API", "response"] if row["name"] == "API" else [label]
        labels = "".join(
            f'<text x="{x + group_width / 2:.1f}" y="{500 if all_tokenizers else 410 + line_index * 12}" '
            f'text-anchor="middle" class="n">{_escape(value)}</text>'
            for line_index, value in enumerate(label_lines)
        )
        groups.append(f'<g>{"".join(bars)}{labels}</g>')
    grids = "".join(
        f'<line x1="52" y1="{y}" x2="860" y2="{y}"/>'
        for y in (top, top + chart_height * 0.25, top + chart_height * 0.5, top + chart_height * 0.75, bottom)
    )
    tick_values = [round(max_tokens * fraction) for fraction in (0, 0.25, 0.5, 0.75, 1)]
    ticks = "".join(
        f'<text x="44" y="{bottom - round(max_tokens * fraction * scale) + 4}" text-anchor="end" class="a">{value}</text>'
        for fraction, value in zip((0, 0.25, 0.5, 0.75, 1), tick_values)
    )
    legend = '<rect x="0" y="0" width="10" height="10" rx="2" fill="#fafafa"/><text x="14" y="9" class="s">AICLTokenizer</text>'
    if any(row.get("gpt4o") is not None for row in rows):
        legend = '<rect x="0" y="0" width="10" height="10" rx="2" fill="#27272a"/><text x="14" y="9" class="s">GPT-4o (optional baseline)</text>' + legend
    subtitle = subtitle.replace("131/131 pass", "lossless checked")
    return f'''<svg width="{width}" height="{height}" xmlns="http://www.w3.org/2000/svg">
<style>.t{{font:700 17px 'Stack Sans Notch','Inter',system-ui,sans-serif;fill:#f9fafb;letter-spacing:-0.3px}}.s{{font:400 11px 'Stack Sans Notch','Inter',system-ui,sans-serif;fill:#9ca3af}}.n{{font:600 10px 'Stack Sans Notch','Inter',system-ui,sans-serif;fill:#e5e7eb}}.a{{font:400 9px 'Stack Sans Notch','Inter',system-ui,sans-serif;fill:#6b7280}}</style>
<rect width="{width}" height="{height}" rx="16" fill="#0a0a0a"/>
<text x="32" y="34" class="t">{_escape(title)}</text>
<text x="32" y="52" class="s">{_escape(subtitle)}</text>
<g transform="translate(32,66)">{legend}</g>
<g stroke="#1f1f23" stroke-width="1">{grids}</g>{ticks}
{''.join(groups)}
<line x1="32" y1="{520 if all_tokenizers else 460}" x2="868" y2="{520 if all_tokenizers else 460}" stroke="#1f1f23"/>
<text x="32" y="{540 if all_tokenizers else 480}" class="s">Lower is better · Stage 1: raw → AICL PUA · Stage 2: BPE → tokens</text>
<text x="32" y="{558 if all_tokenizers else 498}" class="a">AICL benchmark · {len(rows)} tests · stdlib Python</text>
</svg>'''

def self_check() -> None:
    rows = [{"name": "x", "aicl": 2, "gpt3": None, "gpt4": None, "gpt4o": None, "gpt5": None, "llama": None}]
    assert "<svg" in _bar_chart(rows, title="x", subtitle="x", all_tokenizers=False)
    assert "AICL" in _bar_chart(rows, title="x", subtitle="x", all_tokenizers=True)
    print("benchmark self-check: chart generation ok")


def main(argv: Sequence[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--vocab", type=Path, default=ROOT / "tokenizer" / "vocab.json")
    parser.add_argument("--out-dir", type=Path, default=ROOT / "assets")
    parser.add_argument("--baselines", type=Path, help="optional JSON with per-test GPT/LLaMA counts")
    parser.add_argument("--no-svg", action="store_true")
    parser.add_argument("--self-check", action="store_true")
    args = parser.parse_args(argv)
    if args.self_check:
        self_check()
        return 0
    hotpath = None
    if args.vocab.resolve() == (ROOT / "tokenizer" / "vocab.json").resolve():
        hotpath = hotpath_summary()
        print(
            "hotpath 10k: "
            f"cold encode={hotpath['cold_encode_ms']:.1f}ms, "
            f"cold tokenize={hotpath['cold_tokenize_ms']:.1f}ms, "
            f"warm median encode={hotpath['encode_median_ms']:.1f}ms, "
            f"warm median tokenize={hotpath['tokenize_median_ms']:.1f}ms"
        )
    _load, tokenize, _detokenize = _tokenizer_api()
    vocab = _load_vocab(args.vocab)
    rows = measure(vocab, tokenize, load_baselines(args.baselines))
    merge_count, max_length, _version = vocab_summary(vocab)
    for row in rows:
        baseline = ", ".join(f"{key}={row[key]}" for key in ("gpt3", "gpt4", "gpt4o", "gpt5", "llama") if row[key] is not None)
        print(
            f"{row['name']:<16} raw={row['raw']:>3} AICL={row['aicl']:>3}"
            + (f" {baseline}" if baseline else " external-baselines=not-configured")
            + (f" win={row['win']}x" if row["win"] is not None else "")
        )
    print(f"vocab {merge_count} merges maxLen={max_length}; total AICL={sum(row['aicl'] for row in rows)}")
    if not args.no_svg:
        args.out_dir.mkdir(parents=True, exist_ok=True)
        subtitle = f"{merge_count} merges · max {max_length} PUA/token · 8 tests"
        (args.out_dir / "benchmark.svg").write_text(
            _bar_chart(rows, title="AICL vs GPT-4o — tokens (lower is better)", subtitle=subtitle, all_tokenizers=False),
            encoding="utf-8",
        )
        (args.out_dir / "benchmark-all.svg").write_text(
            _bar_chart(rows, title="AICL vs All Tokenizers — 8 tests", subtitle=subtitle, all_tokenizers=True),
            encoding="utf-8",
        )
        print(f"wrote {args.out_dir / 'benchmark.svg'}")
        print(f"wrote {args.out_dir / 'benchmark-all.svg'}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
